### PR TITLE
MTM-53019 add image pull secret for sdk package plugin

### PR DESCRIPTION
**Pull request by AleksanderFiszCumulocity**

### CODE REVIEW

File: microservice/package/src/main/java/com/cumulocity/agent/packaging/BaseMicroserviceMojo.java [hunk 1]
```diff
@@ -121,6 +121,16 @@ public abstract class BaseMicroserviceMojo extends AbstractMojo {
...
     protected String baseImage;
 
+    @Parameter(property = "pull.image.registry")
+    protected String pullRegistry;
+
+    @Parameter(property = "pull.image.username")
+    protected String pullUsername;
+
+    @Parameter(property = "pull.image.password")
+    protected String pullPassword;
```
Comment (kaczynskid-sag) [inline microservice/package/src/main/java/com/cumulocity/agent/packaging/BaseMicroserviceMojo.java]: use the same prefix as `package.docker.baseImage`
```suggestion
    @Parameter(property = "package.docker.registryUrl")
    protected String pullRegistry;

    @Parameter(property = "package.docker.registryUser")
    protected String pullUsername;

    @Parameter(property = "package.docker.registryPass")
    protected String pullPassword;
```

**Review by kaczynskid-sag — COMMENTED** [review]
Isn't the registry part of the `package.docker.baseImage` already?
Can't this credentials be configured in local docker deamon instad?

**Review by ghost — DISMISSED** [review]

**Review by c8y-kh4mell — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

**Review by karol-c8y — DISMISSED** [review]

File: microservice/package/src/main/java/com/cumulocity/agent/packaging/microservice/impl/MicroserviceDockerClientImpl.java [hunk 2]
```diff
@@ -45,9 +45,9 @@ public class MicroserviceDockerClientImpl extends AbstractLogEnabled implements
     private final static ExecutorService executorService = Executors.newSingleThreadExecutor();
 
     @SneakyThrows
-    synchronized public void buildDockerImage(String dockerDirectory, Set<String> tags, Map<String, String> buildArgs, String platform, String networkMode, Integer dockerBuildTimeout) {
+    synchronized public void buildDockerImage(String dockerDirectory, Set<String> tags, Map<String, String> buildArgs, String platform, String networkMode, Integer dockerBuildTimeout, AuthConfigurations pullCOnfiguration) {
```
Comment (c8y-kh4mell) [inline microservice/package/src/main/java/com/cumulocity/agent/packaging/microservice/impl/MicroserviceDockerClientImpl.java]: trivial
```suggestion
    synchronized public void buildDockerImage(String dockerDirectory, Set<String> tags, Map<String, String> buildArgs, String platform, String networkMode, Integer dockerBuildTimeout, AuthConfigurations pullConfiguration) {
```

File: microservice/package/src/test/java/com/cumulocity/agent/packaging/PackageMojoTest.java [hunk 3]
```diff
@@ -228,6 +232,77 @@ public void testDockerBuildSpec() {
...
+
+        AuthConfigurations authConfig = authConfiguration.getValue();
+
+        assertEquals(1, authConfig.getConfigs().size(), "Size of list of auth configurations.");
+        assertTrue(authConfig.getConfigs().containsKey("not.test.registry"), "Contains auth config for registry: not.test.registry");
+    }
+
+    @SneakyThrows
+    @Test
+    public void pullRegistryShouldBeCalculatedBaseImageIfEmpty(){
```
Comment (c8y-kh4mell) [inline microservice/package/src/test/java/com/cumulocity/agent/packaging/PackageMojoTest.java]:
```suggestion
    public void pullRegistryShouldCalculateBaseImageIfEmpty(){
```

**Review by c8y-kh4mell — DISMISSED** [review]

**Review by c8y-kh4mell — DISMISSED** [review]

File: microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java [hunk 4]
```diff
@@ -234,7 +238,14 @@ private void buildDockerImage(String targetBuildArch) {
             log.warn("Your image might be incompatible with cloud hosting!");
         }
 
-        dockerClient.buildDockerImage(dockerWorkDir.getAbsolutePath(),tags,buildArgs, targetBuildArch, dockerBuildNetwork, dockerBuildTimeout);
+       AuthConfigurations authConfigurations = new AuthConfigurations();
+        if(isNotBlank(pullUsername) && isNotBlank(pullPassword)){
```
Comment (jkot-c8y) [inline microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java]: minor: maybe log.info() with some details in this IF branch?

**Review by jkot-c8y — DISMISSED** [review]

**Review by jkot-c8y — DISMISSED** [review]

**Review by karol-c8y — DISMISSED** [review]

File: microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java [hunk 5]
```diff
@@ -432,5 +434,64 @@ private static class ManifestConstraintViolation {
...
+                missingValues.add("registry");
+            }
+
+            if (isBlank(user)) {
+                missingValues.add("Username");
+            }
+
+            if (isBlank(password)) {
+                missingValues.add("Password");
+            }
```
Comment (kaczynskid-sag) [inline microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java]: Should correspond to existing available settings

File: microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java [hunk 6]
```diff
@@ -432,5 +434,64 @@ private static class ManifestConstraintViolation {
...
+        private boolean displayNoPullCredentialInfo(){
+            return isNotBlank(getRegistry()) || isNotBlank(user) || isNotBlank(password);
+        }
+
+        private String getRegistry(){
+            if (isNotBlank(registry)) {
+                return registry;
+            }
+
+            return baseImage.contains("/") ? baseImage.substring(0, baseImage.lastIndexOf("/")) : null;
```
Comment (kaczynskid-sag) [inline microservice/package/src/main/java/com/cumulocity/agent/packaging/PackageMojo.java]: default is docker-hub, not null

**Review by kaczynskid-sag — CHANGES_REQUESTED** [review]

**Review by kaczynskid-sag — APPROVED** [review]

**Review by karol-c8y — APPROVED** [review]

**Pull request merged**